packages/es-components/src/components/controls/buttons/DropdownButton.js

```diff
@@ -40,15 +45,89 @@ const StyledButtonLink = styled(Button)`
   }
 `;
 
+const TAB_KEY_CODE = 9;
+const UP_ARROW_CODE = 38;
+const DOWN_ARROW_CODE = 40;
+
+function getFocusables(node) {
+  const focusableElements = node.querySelectorAll('button');
+  const firstFocusable = focusableElements[0];
+  const lastFocusable = focusableElements[focusableElements.length - 1];
+
+  return {
+    focusableElements,
+    firstFocusable,
+    lastFocusable
+  };
+}
+
+function isCurrentlyActive(node) {
+  return document.activeElement === node;
+}
+
+function focusTrap(node) {
+  const { firstFocusable, lastFocusable } = getFocusables(node);
+
+  node.addEventListener('keydown', event => {
+    if (event.key === 'Tab' || event.keyCode === TAB_KEY_CODE) {
+      if (event.shiftKey) {
+        if (isCurrentlyActive(firstFocusable)) {
+          lastFocusable.focus();
+          event.preventDefault();
+        }
+      } else if (isCurrentlyActive(lastFocusable)) {
+        firstFocusable.focus();
+        event.preventDefault();
+      }
+    }
+  });
+}
+
+function arrowMovement(node) {
+  const { focusableElements, firstFocusable, lastFocusable } = getFocusables(
+    node
+  );
+  const focusables = [...focusableElements];
+
+  node.addEventListener('keydown', event => {
```

Comment:
The only thing slightly bothering me with this is how well the events get cleaned up when the `DropdownButton` is unmounted. I don't think this is severe, but if we add these dropdowns to something like a modal I can see these event listeners piling up in memory.

If cleanup is a concern, lightweight option if we want to clean up these event listeners when the nodes are going to be removed from the DOM could be to return a `removeEventListener` thunk from these functions to call in `componentWillUnmount`. 

```js
function focusTrap(node) {
  const { firstFocusable, lastFocusable } = getFocusables(node);

  const eventListener = event => {
    // ...
  };

  node.addEventListener('keydown', eventListener);

  return () => node.removeEventListener('keydown', eventListener);
}
```

```js
componentDidMount() {
  this.clearFocusTrap = focusTrap(this.buttonDropdown);
  this.clearArrowMovement = arrowMovement(this.buttonDropdown);
}

componentWillUnmount() {
  this.clearFocusTrap();
  this.clearArrowMovement();
}
```

Reply:
That's a fair point. I will add this. 👍